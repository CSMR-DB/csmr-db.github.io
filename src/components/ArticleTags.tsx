import React from 'react'
import { StyledLink } from './StyledLink'
import styled, { StyledComponent } from 'styled-components'

const StyledP: StyledComponent<'p', any, {}, never> = styled.p`
  text-align: left !important;
`

export const ArticleTags: ({ tags }: { tags?: string[] }) => JSX.Element = ({
  tags,
}: {
  tags?: string[]
}): JSX.Element => (
  <StyledP>
    Tags:{' '}
    {tags &&
      tags.map(
        (tag: string, i: number): JSX.Element => (
          <span key={i}>
            <StyledLink
              to={`/tag/${tag.toLocaleLowerCase().replace(/\s/gi, '_')}`}
            >
              {tag}
            </StyledLink>
            {i + 1 < tags.length ? ', ' : ''}
          </span>
        )
      )}
  </StyledP>
)
