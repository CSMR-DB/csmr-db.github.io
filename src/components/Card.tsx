import styled, { StyledComponent } from 'styled-components'
// tslint:disable-next-line: no-implicit-dependencies
import { BackgroundColorProperty } from 'csstype'

export const Card: StyledComponent<'article', any, {}, never> = styled.article`
  position: relative;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  background: white;
  box-shadow: 0 0.075rem 0.075rem rgba(0, 0, 0, 0.1),
    0 0.125rem 0.125rem rgba(0, 0, 0, 0.25);
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);

  &:hover {
    box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.25),
      0 0.25rem 0.25rem rgba(0, 0, 0, 0.2);
    transform: translateY(-0.25rem);
  }

  & h1 {
    font-size: 1.5em;
    margin: 0;
  }

  & h2 {
    font-size: 1.1em;
    font-style: italic;
    margin: 0;
    color: #aaa;
    margin-bottom: 1rem;
  }

  & p {
    margin: 0;
    text-align: justify;
  }

  & header {
    margin-bottom: 2em;
  }
`

export const CardHeader: StyledComponent<
  'header',
  any,
  {},
  never
> = styled.header`
  margin-bottom: 2em;
  max-height: 18rem;
  overflow: hidden;
  grid-area: 'header';
`

export const CardHeaderText: StyledComponent<
  'div',
  any,
  {},
  never
> = styled.div`
  padding-left: 2em;
`

interface ICardHeaderIconWrapperProps {
  backgroundColor: BackgroundColorProperty
}

export const CardHeaderIconWrapper: StyledComponent<
  'div',
  any,
  ICardHeaderIconWrapperProps,
  never
> = styled.div`
  border-radius: 999px;
  overflow: hidden;
  max-width: 4em;
  width: 100%;
  max-height: 4em;
  height: 100%;
  background: ${(props: ICardHeaderIconWrapperProps): BackgroundColorProperty =>
    props.backgroundColor || '#DDD'};
  padding: 0.5em;
`

export const CardBody: StyledComponent<'main', any, {}, never> = styled.main`
  position: relative;
  padding: 2em;
  grid-area: 'body';
`

export const CardFooter: StyledComponent<
  'footer',
  any,
  {},
  never
> = styled.footer`
  background: #eee;
  padding: 2em;
  border-top: 1px solid grey;
  grid-area: 'footer';

  & span {
    padding: 0 0.25em;
  }
`
