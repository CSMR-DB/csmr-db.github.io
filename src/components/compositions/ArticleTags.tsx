import React from 'react'
import styled, { StyledComponent } from 'styled-components'

import { StyledLink } from '../StyledLink'

const StyledArticleTags: StyledComponent<'p', any, {}, never> = styled.p`
  text-align: left !important;
`

interface IArticleTagsProps {
  tags?: string[]
}

export function ArticleTags({ tags }: IArticleTagsProps): JSX.Element {
  return (
    <StyledArticleTags>
      Tags:{' '}
      {tags &&
        tags.map(
          (tag: string, i: number): JSX.Element => (
            <span key={i}>
              <StyledLink
                to={`/skillset/${tag.toLocaleLowerCase().replace(/\s/gi, '_')}`}
                $_testid={'SkillPageLink'}
              >
                {tag}
              </StyledLink>
              {i + 1 < tags.length ? ', ' : ''}
            </span>
          )
        )}
    </StyledArticleTags>
  )
}
