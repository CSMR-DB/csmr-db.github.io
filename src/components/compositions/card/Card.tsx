import React, { ReactNode } from 'react'
import styled, {
  StyledComponent,
  DefaultTheme,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { Tween } from 'react-gsap'

export interface ICardProps {
  theme?: DefaultTheme
  children: ReactNode
  index?: number
}

export const StyledCard: StyledComponent<
  'article',
  any,
  ICardProps,
  never
> = styled.article`
  position: relative;
  height: 100%;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  box-shadow: 0 0.075rem 0.075rem rgba(0, 0, 0, 0.1),
    0 0.125rem 0.125rem rgba(0, 0, 0, 0.25);
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);

  & header img {
    transition: all 0.2s ease-in-out !important;
  }

  & h1 {
    font-size: 1.1rem;
    margin: 0;
  }

  & h2 {
    font-size: 1rem;
    font-style: italic;
    margin: 0;
    color: #aaa;
    margin-bottom: 1rem;
  }

  & p {
    text-align: left;
  }

  ${({ theme }: ICardProps): FlattenSimpleInterpolation => css`
    background: ${theme!.palette.light.normal};

    @media ${theme!.breakpoints.min.smartphone} {
      text-align: justify;
    }

    @media ${theme!.breakpoints.min.desktop} {
      &:hover {
        box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.25),
          0 0.25rem 0.25rem rgba(0, 0, 0, 0.2);

        & header img {
          transform: scale(1.1);
        }
      }
    }
  `}
`

export function Card({ children, index = 0 }: ICardProps): JSX.Element {
  return (
    <Tween
      from={{ scale: 0.95, opacity: 0 }}
      duration={0.5}
      delay={0.25 + index * 0.125}
      ease="back"
    >
      <StyledCard>{children}</StyledCard>
    </Tween>
  )
}
