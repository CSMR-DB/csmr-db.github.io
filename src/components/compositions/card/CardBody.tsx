import styled, { StyledComponent } from 'styled-components'

export const CardBody: StyledComponent<'main', any, {}, never> = styled.main`
  position: relative;
  padding: 2em;
  grid-area: 'body';
`
