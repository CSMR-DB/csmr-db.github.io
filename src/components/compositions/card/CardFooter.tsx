import styled, {
  StyledComponent,
  FlattenSimpleInterpolation,
  css,
} from 'styled-components'

interface ICardFooterProps {
  $style?: boolean
}

export const CardFooter: StyledComponent<
  'footer',
  any,
  ICardFooterProps,
  never
> = styled.footer`
  ${({ $style: style = false }: ICardFooterProps): FlattenSimpleInterpolation =>
    style
      ? css`
          background: #eee;
          padding: 2rem;
          border-top: 1px solid #ddd;
          grid-area: footer;
        `
      : css`
          padding: 0 2rem 2rem;
        `}
`
