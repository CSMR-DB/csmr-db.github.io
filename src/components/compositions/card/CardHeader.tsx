import styled, { StyledComponent } from 'styled-components'

export const CardHeader: StyledComponent<
  'header',
  any,
  {},
  never
> = styled.header`
  grid-area: 'header';
`
