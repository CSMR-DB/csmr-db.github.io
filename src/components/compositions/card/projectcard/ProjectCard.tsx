import React from 'react'
import Image from 'gatsby-image'

import { FileChildImageSharp } from '../../../../types/graphql.types'

import { Card } from '../Card'
import { ArticleTags } from '../../../compositions/ArticleTags'
import { StyledLink } from '../../../StyledLink'
import { DateFormatted } from '../../../DateFormatted'
import { CardHeader } from '../CardHeader'
import { CardFooter } from '../CardFooter'
import { CardBody } from '../CardBody'
import { CardHeaderText } from '../CardHeaderText'

interface IProjectCardProps {
  title: string
  body: string
  tags: string[]
  date: string
  generatedRoute: string
  image?: FileChildImageSharp
  index?: number
}

export function ProjectCard({
  title,
  body,
  tags,
  date,
  generatedRoute,
  image,
  index = 0,
}: IProjectCardProps): JSX.Element {
  return (
    <Card index={index} $_testid={'ProjectCard'}>
      <CardHeader>
        {image && <Image fluid={image.childImageSharp.fluid} />}
        <CardHeaderText>
          <h1>
            <StyledLink to={generatedRoute} $_testid={'ProjectPageLink'}>
              {title}
            </StyledLink>
          </h1>
          <h6>
            <DateFormatted dateString={date} />
          </h6>
        </CardHeaderText>
      </CardHeader>
      <CardBody>
        <p>{body}</p>
      </CardBody>
      <CardFooter>
        <ArticleTags tags={tags} />
      </CardFooter>
    </Card>
  )
}
