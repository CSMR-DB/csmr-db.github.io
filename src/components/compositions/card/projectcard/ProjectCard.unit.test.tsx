import React, { Component } from 'react'
import { ThemeProvider } from 'styled-components'
import { create, ReactTestRenderer } from 'react-test-renderer'
import { ReactWrapper, mount } from 'enzyme'

import { StaticDataManager } from '../../../../data/StaticDataManager'

import { ProjectCard } from './ProjectCard'
import { Card } from '../Card'
import { CardHeader } from '../CardHeader'
import { StyledLink } from '../../../StyledLink'
import { DateFormatted } from '../../../DateFormatted'
import { CardBody } from '../CardBody'
import { CardFooter } from '../CardFooter'
import { ArticleTags } from '../../ArticleTags'

describe('ProjectCard', (): void => {
  it('should match snapshot', (): void => {
    const comp: ReactTestRenderer = create(
      <ThemeProvider theme={StaticDataManager.theme}>
        <ProjectCard
          title={'Jest a Project'}
          body={'What more is there to say, really?'}
          generatedRoute={'/projects/jest_a_project'}
          date={'1994-01-08T00:00:00+0000'}
          tags={['Jest', 'Enzyme']}
        />
      </ThemeProvider>
    )

    expect(comp).toMatchSnapshot()
  })

  it('should render internally', (): void => {
    const comp: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>> = mount(
      <ThemeProvider theme={StaticDataManager.theme}>
        <ProjectCard
          title={'Jest a Project'}
          body={'What more is there to say, really?'}
          generatedRoute={'/projects/jest_a_project'}
          date={'1994-01-08T00:00:00+0000'}
          tags={['Jest', 'Enzyme']}
        />
      </ThemeProvider>
    )

    expect(comp.find(Card).length).toEqual(1)

    expect(
      comp.find(Card).find(CardHeader).find('h1').find(StyledLink).length
    ).toEqual(1)
    expect(
      comp.find(Card).find(CardHeader).find('h1').find(StyledLink).text()
    ).toEqual('Jest a Project')
    expect(
      comp.find(Card).find(CardHeader).find('h1').find(StyledLink).props().to
    ).toEqual('/projects/jest_a_project')
    expect(
      comp.find(Card).find(CardHeader).find('h6').find(DateFormatted).length
    ).toEqual(1)
    expect(
      comp.find(Card).find(CardHeader).find('h6').find(DateFormatted).props()
        .dateString
    ).toEqual('1994-01-08T00:00:00+0000')
    expect(
      comp.find(Card).find(CardHeader).find('h6').find(DateFormatted).text()
    ).toEqual('08 Jan 1994')

    expect(comp.find(Card).find(CardBody).length).toEqual(1)
    expect(comp.find(Card).find(CardBody).text()).toContain(
      'What more is there to say, really?'
    )

    expect(comp.find(Card).find(CardFooter).find(ArticleTags).length).toEqual(1)
    expect(
      comp.find(Card).find(CardFooter).find(ArticleTags).props().tags!.length
    ).toEqual(2)
    expect(comp.find(Card).find(CardFooter).find(ArticleTags).text()).toEqual(
      'Tags: Jest, Enzyme'
    )
  })
})
