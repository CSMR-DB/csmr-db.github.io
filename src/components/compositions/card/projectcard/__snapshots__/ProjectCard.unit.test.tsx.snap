// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard should match snapshot 1`] = `
.c0 {
  position: relative;
  height: 100%;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  box-shadow: 0 0.075rem 0.075rem rgba(0,0,0,0.1), 0 0.125rem 0.125rem rgba(0,0,0,0.25);
  -webkit-transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  background: #fff;
}

.c0 header img {
  -webkit-transition: all 0.2s ease-in-out !important;
  transition: all 0.2s ease-in-out !important;
}

.c0 h1 {
  font-size: 1.1rem;
  margin: 0;
}

.c0 h2 {
  font-size: 1rem;
  font-style: italic;
  margin: 0;
  color: #aaa;
  margin-bottom: 1rem;
}

.c0 p {
  text-align: left;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
  color: #f00 !important;
}

.c3:hover {
  color: #d00;
}

.c3.active {
  color: #b00;
}

.c6 {
  text-align: left !important;
}

.c1 {
  grid-area: 'header';
}

.c5 {
  grid-area: 'footer';
  padding: 0 2rem 2rem;
}

.c4 {
  grid-area: 'body';
  position: relative;
  padding: 2rem;
}

.c2 {
  padding: 2rem 2rem 0 2rem;
}

.c2 > h6 {
  margin-bottom: 0;
}

@media only screen and (min-width:600px) {
  .c0 {
    text-align: justify;
  }
}

@media only screen and (min-width:1200px) {
  .c0:hover {
    box-shadow: 0 0.125rem 0.25rem rgba(0,0,0,0.25), 0 0.25rem 0.25rem rgba(0,0,0,0.2);
  }

  .c0:hover header img {
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
  }
}

<article
  className="c0"
>
  <header
    className="c1"
  >
    <div
      className="c2"
    >
      <h1>
        <a
          className="c3"
          href="/projects/jest_a_project"
        >
          Jest a Project
        </a>
      </h1>
      <h6>
        <span>
          08 Jan 1994
        </span>
      </h6>
    </div>
  </header>
  <main
    className="c4"
  >
    <p>
      What more is there to say, really?
    </p>
  </main>
  <footer
    className="c5"
  >
    <p
      className="c6"
    >
      Tags:
       
      <span>
        <a
          className="c3"
          href="/skillset/jest"
        >
          Jest
        </a>
        , 
      </span>
      <span>
        <a
          className="c3"
          href="/skillset/enzyme"
        >
          Enzyme
        </a>
        
      </span>
    </p>
  </footer>
</article>
`;
